code/modules/mob/living/silicon/robot/robot_model.dm

```diff
@@ -350,14 +348,205 @@
+
+		if(!do_after(robot_owner, 4 SECONDS, interaction_key = "auto_wash_toggle", extra_checks = allow_buffer_change))
+			wash_audio.stop() // Coward
+			animate(robot_owner, pixel_x = base_x, pixel_y = base_y, time = 1)
+			return FALSE
+	else
+		if(!COOLDOWN_FINISHED(src, toggle_cooldown))
+			robot_owner.balloon_alert(robot_owner, "auto-wash deactivating, please hold...")
+			return FALSE
+		robot_owner.balloon_alert(robot_owner, "de-activating auto-wash...")
+
+	toggle_wash()
+
+/datum/action/toggle_buffer/proc/toggle_wash()
+	if(buffer_on)
+		deactivate_wash()
+	else
+		activate_wash()
+
+/datum/action/toggle_buffer/proc/activate_wash()
+	var/mob/living/silicon/robot/robot_owner = owner
+	buffer_on = TRUE
+	// Slow em down a bunch
+	robot_owner.add_movespeed_modifier(/datum/movespeed_modifier/auto_wash)
+	RegisterSignal(robot_owner, COMSIG_MOVABLE_MOVED, .proc/clean)
+	//This is basically just about adding a shake to the borg, effect should look ilke an engine's running
+	var/base_x = robot_owner.base_pixel_x
+	var/base_y = robot_owner.base_pixel_y
+	robot_owner.pixel_x = base_x + rand(-7, 7)
+	robot_owner.pixel_y = base_y + rand(-7, 7)
+	//Larger shake with more changes to start out, feels like "Revving"
+	animate(robot_owner, pixel_x = base_x, pixel_y = base_y, time = 1, loop = -1)
+	for(var/i in 1 to 100)
+		var/x_offset = base_x + rand(-2, 2)
+		var/y_offset = base_y + rand(-2, 2)
+		animate(pixel_x = x_offset, pixel_y = y_offset, time = 1)
+	if(!wash_audio.is_active())
+		wash_audio.start()
+	clean()
+	UpdateButtonIcon()
+
+/datum/action/toggle_buffer/proc/deactivate_wash()
+	var/mob/living/silicon/robot/robot_owner = owner
+	var/time_left = timeleft(wash_audio.timerid) // We delay by the timer of our wash cause well, we want to hear the ramp down
+	var/finished_by = time_left + 2.6 SECONDS
+	// Need to ensure that people don't spawn the deactivate button
+	COOLDOWN_START(src, toggle_cooldown, finished_by)
+	// Diable the cleaning, we're revving down
+	UnregisterSignal(robot_owner, COMSIG_MOVABLE_MOVED)
+	// Do the rumble animation till we're all finished
+	var/base_x = robot_owner.base_pixel_x
+	var/base_y = robot_owner.base_pixel_y
+	animate(robot_owner, pixel_x = base_x, pixel_y = base_y, time = 1)
+	for(var/i in 1 to finished_by - 0.1 SECONDS) //We rumble until we're finished making noise
+		var/x_offset = base_x + rand(-1, 1)
+		var/y_offset = base_y + rand(-1, 1)
+		animate(pixel_x = x_offset, pixel_y = y_offset, time = 1)
+	// Reset our animations
+	animate(pixel_x = base_x, pixel_y = base_y, time = 2)
+	addtimer(CALLBACK(wash_audio, /datum/looping_sound/proc/stop), time_left)
+	addtimer(CALLBACK(src, .proc/turn_off_wash), finished_by)
+
+/datum/action/toggle_buffer/proc/turn_off_wash()
+	var/mob/living/silicon/robot/robot_owner = owner
+	buffer_on = FALSE
+	robot_owner.remove_movespeed_modifier(/datum/movespeed_modifier/auto_wash)
+	UpdateButtonIcon()
+
+/datum/action/toggle_buffer/proc/allow_buffer_change()
```

Comment:
Some of these procs should be documented

Reply:
Sure